packages/codelabs/src/basics/lit-html.md

```diff
@@ -0,0 +1,1139 @@
+
+    <script type="module">
+      ...code sample goes here...
+    </script>
+  </body>
+</html>
+```
+
+</detail>
+
+## Adding a todo
+
+Now, we will add the ability to add todos to our list.
+
+Start by adding an input field and a button:
+
+```js
+class TodoApp extends LitElement {
+  constructor() {
+    super();
+    this.todos = [
+      { text: 'Do A', finished: true },
+      { text: 'Do B', finished: false },
+      { text: 'Do C', finished: false }
+    ];
+  }
+
+  render() {
+    return html`
+      <h1>Todo app</h1>
+
+      <input id="addTodoInput" placeholder="Name">
+      <button @click=${this._addTodo}>Add</button>
+
+      <ol>
+      ${this.todos.map(todo => html`
+        <li>${todo.text} (${todo.finished ? 'Finished' : 'Unfinished')})</li>
+      `)}
+      </ol>
+
+      ${footerTemplate}
+    `;
+  }
+}
+```
+
+On the add button, we added an event listener for the `click` event. This is done by prefixing the event name with a `@`:
+
+```js
+<button @click=${this._addTodo}>
+```
+
+The value of an event listener should be a function. In this case, we reference a method on our component, which we should now implement:
+
+```js
+_addTodo() {
+  const input = this.shadowRoot.getElementById('addTodoInput');
+  const text = input.value;
+  input.value = '';
+
+  this.todos.push([{ text, finished: false }]);
+  this.requestUpdate();
+}
+```
+
+When this event handler is called, we create a new todo item and add it to the array of todos. Because we're mutating the todos array, `LitElement` is not able to pick up that we changed something. We can let `LitElement` know about this by calling the `requestUpdate` which exists on any element that extends from `LitElement`.
+
+When you click add, you should see the new element appear on the screen.
+
+This allows us to observe the power of lit-html in action. If you inspect the DOM while adding a todo item, you will see that only the new todo item is actually flashing:
+
+TODO: add img
+
+![todo](./assets/foo.png)
+
+When something in the DOM inspector flashes, it means that the browser is doing actual work to update the DOM tree. This is very expensive work, things like styles and layout need to be recalculated up and down the tree, so you want to minimize this as much as possible. Due to the way lit-html templating works it knows precisely what changed where and it will update only that part, without requiring a lot of from the developer.
+
+<detail>
+  <summary>View final result</summary>
+
+```html
+<!DOCTYPE html>
+<html>
+  <head></head>
+
+  <body>
+    <todo-app></todo-app>
+
+    <script type="module">
+      ...code sample goes here...
+    </script>
+  </body>
+</html>
+```
+
+</detail>
+
+## Reactive property changes
+
+Right now we're triggering an update manually when we make a change. This is fine for some use cases, but it can get pretty cumbersome to do this manually all the time and you will not be able to respond updated triggered by a parent element.
```

Comment:
```suggestion
Right now we're triggering an update manually when we make a change. This is fine for some use cases, but it can get pretty cumbersome to do this manually all the time and you will not be able to respond to an update triggered by a parent element.
```